[FLINK-36123][table] Add the built-in function PERCENTILE

dylanhz opened the pull request:
## What is the purpose of the change

Add the built-in function PERCENTILE.
Examples:
```SQL
> SELECT PERCENTILE(col, 0.3) FROM VALUES (0), (10), (10) AS tab(col);
 6.0

> SELECT PERCENTILE(col, 0.3, freq) FROM VALUES (0, 1), (10, 2) AS tab(col, freq);
 6.0

> SELECT PERCENTILE(col, array(0.25, 0.75)) FROM VALUES (0), (10) AS tab(col);
 [2.5,7.5]
```

## Brief change log

- Add `ArgumentTypeStrategy` for percentage (array) value with literal value checking
- Refactor error test for built-in agg functions
- Support creating `AggregateInfo` through `BuiltInFunctionDefinition` directly to avoid an extra `SqlAggFunction` conversion
- [FLINK-36123](https://issues.apache.org/jira/browse/FLINK-36123)

## Verifying this change

For the two `ArgumentTypestrategy`: `InputTypeStrategiesTest`

For the `PercentileTypestrategy`: `TypeStrategiesTest`

For the function PERCENTILE:
- `PercentileAggFunctionITCase`,
- `WindowAggregateITCase#testPercentileOnEventTimeTumbleWindow()`
- `stream/OverAggregateITCase#testPercentile()`, `batch/OverAggregateITCase#testPercentile()`
- `sql/AggregateITCase#testSinglePercentile()`, `sql/AggregateITCase#testMultiPercentile()`
- `table/AggregateITCase#testSinglePercentile()`, `table/AggregateITCase#testMultiPercentile()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

lincoln-lil commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/PercentageArgumentTypeStrategy.java:
```diff
@@ -0,0 +1,80 @@
+
+import java.util.Optional;
+
+/** An {@link ArgumentTypeStrategy} that expects a percentage value between [0.0, 1.0]. */
```
Do we have precision requirements?

dylanhz replied:
This strategy will only perform value checking when the input is a literal value, I believe that precision requirements are unnecessary because the boundaries of 0.0 and 1.0 do not experience any loss of precision.

lincoln-lil replied:
Behind this question, I'd like to say, for percentage representations, I think it makes a lot of sense to use exact numbers rather than approximate numbers(https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/types/#exact-numerics), and there is no need to support excessive precision, WDYT?

lincoln-lil replied:
I also checked other systems like pg and it looks like we don't have to add restrictions, so I'm fine with the current implementation

lincoln-lil commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/PercentageArgumentTypeStrategy.java:
```diff
@@ -0,0 +1,80 @@
+
+    private final boolean expectedNullability;
+
+    public PercentageArgumentTypeStrategy(boolean isNullable) {
```
nit: `isNullable` -> `nullable`

lincoln-lil commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/PercentageArgumentTypeStrategy.java:
```diff
@@ -0,0 +1,80 @@
+@Internal
+public final class PercentageArgumentTypeStrategy implements ArgumentTypeStrategy {
+
+    private final boolean expectedNullability;
```
`requireNotNull` might be more straightforward

dylanhz replied:
The term `expected...` aligns with the naming conventions used in other strategies, such as `FamilyArgumentTypeStrategy` and `RootArgumentTypeStrategy`. It might be better to maintain this consistency.

lincoln-lil replied:
make sense

lincoln-lil commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java:
```diff
@@ -125,6 +125,19 @@ public static InputTypeStrategy windowTimeIndicator() {
+     * An {@link ArgumentTypeStrategy} that expects an array of percentages with each element
+     * between [0.0, 1.0].
+     */
+    public static ArgumentTypeStrategy percentageArray(boolean expectedNullability) {
```
Is it possible to replace this new strategy with a `and` strategy that combines `ArrayTypeStrategy` and `percentage`?

dylanhz replied:
I think `and` strategy describes a relationship like `a + b`, but here we need a inclusion relationship which has not been supported yet. 
For example, if we want a strategy like `ARRAY<NUMERIC>`, then we have to create a new strategy class specially or in a very complex way, like `or(explicit(ARRAY(DOUBLE())), explicit(ARRAY(FLOAT()))...)`. This can be an improvement in the future.
Back to our implementation, we have to perform literal value checking additionally, so I took the compromised and easy way which is creating a new strategy specially.

lincoln-lil replied:
Reasonable choice : )

lincoln-lil reviewed (commented):
@dylanhz Thanks for working on this!  I've left some comment for the 1st commit. Will continue reviewing the subsequent commits.

lincoln-lil commented on docs/data/sql_functions.yml:
```diff
@@ -251,6 +251,21 @@ arithmetic:
+      `(NUMERIC: INTEGER_NUMERIC, FLOAT, DOUBLE, DECIMAL)`
+      
+      Returns a `DOUBLE` if percentage is numeric, or an `ARRAY<DOUBLE>` if percentage is an array. `NULL` if percentage is an empty array.
 
```
It would be good to add some hints about performance, such as suggesting that the user consider applying this agg func in a window scenario (which usually yields better performance), whereas in a normal group agg scenario, the performance overhead of a full sort triggered by each record is something the user needs to be concerned about.

lincoln-lil replied:
This should be for the last commit.

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInAggregateFunctionTestBase.java:
```diff
@@ -362,6 +387,70 @@ public void execute(TableEnvironment tEnv, Table sourceTable) {
+        private final String exceptionMessage;
+
+        public ValidationErrorItem(String exceptionMessage) {
+            this.exceptionClass = ValidationException.class;
```
We can parameterize this `exceptionClass` and change `ValidationErrorItem` to `ErrorTestItem` for a more general purpose, otherwise a constant is enough.

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInAggregateFunctionTestBase.java:
```diff
@@ -362,6 +387,70 @@ public void execute(TableEnvironment tEnv, Table sourceTable) {
+        protected abstract void getResult(TableEnvironment tEnv, Table sourceTable);
+    }
+
+    private static final class SqlValidationErrorItem extends ValidationErrorItem {
```
`SqlValidationErrorItem` -> `SqlErrorTestItem`

lincoln-lil commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInAggregateFunctionTestBase.java:
```diff
@@ -362,6 +387,70 @@ public void execute(TableEnvironment tEnv, Table sourceTable) {
+        }
+    }
+
+    private static final class TableApiValidationErrorItem extends ValidationErrorItem {
```
`TableApiValidationErrorItem` -> `TableApiErrorTestItem`

lincoln-lil reviewed (commented):
Commets for the 2nd commit.

lincoln-lil commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala:
```diff
@@ -501,9 +497,17 @@ object AggregateUtil extends Enumeration {
+      // BuiltInFunctionDefinition to SqlAggFunction is unnecessary, we can simply create
+      // AggregateInfo through BuiltInFunctionDefinition and runtime implementation (obtained from
+      // AggFunctionFactory) directly.
+      // NOTE: make sure to use .runtimeProvided() in BuiltInFunctionDefinition in this case.
```
What'll happen if not use `.runtimeProvided()`?

lincoln-lil replied:
Should add a check if an error occurs.

dylanhz replied:
if `.runtimeProvided()` and `.runtimeClass()` both unused and without a conversion rule, the automatic conversion from `BuiltInFunctionDefinition` to `BridgingSqlAggFunction` will fail because implementation not found, and eventually throw an error, this checking logic is in `FunctionCatalogOperatorTable#verifyFunctionKind()`.

For those agg functions with a conversion rule, this is ok because even though the above process fails, they can still create `SqlAggFunction` through the conversion rule. Also, these functions are supposed to use `.runtimeDeferred()` as a mark of using conversion rule, but only a few them actually did that.

BTW, this implementation checking will not be conducted in Table API which may be a potential problem. In fact, if you delete `.runtimeProvided()` in definition, calling function in Table API can still work while SQL can't.

lincoln-lil replied:
So if someone adds a new agg func that is declared in the wrong way, it will not pass the test (the test case should cover both sql and table api).
In that case I have no problem.

lincoln-lil reviewed (commented):
Comments for the 3rd commit

lincoln-lil commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java:
```diff
@@ -0,0 +1,382 @@
+                        : DecimalDataUtils.doubleValue((DecimalData) value);
+
+        long cnt = Optional.ofNullable(acc.valueCount.get(val)).orElse(0L) + frequency;
+        if (cnt < 0) {
```
We can't simpliy ignore this case, should update the count into state because there may exists non-sequential update inputs.

lincoln-lil replied:
Similar logic can be found in `MaxWithRetractAccumulator`

lincoln-lil commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java:
```diff
@@ -0,0 +1,382 @@
+        }
+
+        public Double[] getValue() {
+            List<Pair<Double, Integer>> sortedPercentages =
```
Avoid Java Streams (https://flink.apache.org/zh/how-to-contribute/code-style-and-quality-java/#java-streams)

lincoln-lil reviewed (commented):
Comments for the last commit

lincoln-lil commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java:
```diff
@@ -0,0 +1,382 @@
+                long higher = (long) Math.ceil(position);
+
+                while (preCnt < lower) {
+                    j++;
```
Should we prevent `j` out of bound of the `sortedList`?

dylanhz replied:
You are right, I forgot that `lower` can be `totalCount + 1` when `percentage = 1.0`.

lincoln-lil reviewed (commented):
Added a few comments on the `PercentileAggFunction` implementation.
We need to add tests into `OverAggITCase` and also add tests for batch scenarios.

dylanhz commented on flink-python/pyflink/table/tests/test_expression.py:
```diff
@@ -126,6 +126,8 @@ def test_expression(self):
         self.assertEqual("UNHEX(a)", str(expr1.unhex))
         self.assertEqual('truncate(a, 3)', str(expr1.truncate(3)))
+        self.assertEqual('PERCENTILE(a, 0.5)', str(expr1.percentile(0.5)))
+        self.assertEqual('PERCENTILE(a, 0.5, b)', str(expr1.percentile(0.5, expr2)))
```
What's the diff between a literal value and an array? I mean they are both expressions, and this test doesn't care about actual argument types.

lincoln-lil replied:
To ensure the functionality which the user passes a percentage array works fine(as the sql tests do).

dylanhz commented:
@lincoln-lil  Thank you for the feedback. I have fixed all the issues. 

BTW, there were some problems with the original position calculation method; it is now fixed and consistent with the definition of Spark. 

Please check the latest commit when you have time, especially the function implementation part.

lincoln-lil commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java:
```diff
@@ -0,0 +1,383 @@
+                long lower = (long) Math.floor(position);
+                long higher = (long) Math.ceil(position);
+
+                // lower <= (totalCount - 1) * percentage + 1 <= totalCount
```
This holds true when state ttl is not set.
However, once ttl is set, the `totalCount` and `valueCount` states in acc are not atomic, and there may be corner cases where `totalCount` still exists but the `valueCount` state has expired, in which case the `totalCount` value may be much larger than the actual state of `valueCount`

dylanhz replied:
`totalCount` has been removed from accumulator in the latest commit, and will be calculated in `getValue()` once needed.

lincoln-lil commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java:
```diff
@@ -75,35 +74,45 @@ public boolean equals(Object o) {
-            }
-            sortedPercentages.sort(Comparator.comparing(Pair::getKey));
+            // calculate totalCount here to avoid the inconsistencies between valueCount and
+            // totalCount, which may arise if both are stored in the accumulator and have different
```
nit: It's not so much the different state ttl settings, it's the fact that operating serially on two different states (ValueState acc.totalCount vs MapState acc.valueCount) can't achieve consistent expiration

lincoln-lil reviewed (approved):
@dylanhz Thanks for your updates! This pr is good for merging now except one minor comment. Could you also reorg the commits before the merge?
